### PR TITLE
Identity aware GetNodes.

### PR DESCRIPTION
**Pull request by russjones**
**Purpose**

Make `GetNodes` identity aware to only return servers that the caller has access to. For service accounts (like `RoleProxy` or `RoleAdmin`) return all nodes.

**Implementation**

* Make logging in `role.go` much more verbose, but only log while in debug mode.
* When returning a list of nodes, check if the callers `services.RoleSet` has access to it, and only include it if it has access.
* Make the Web UI used the credentials of the logged in user to request a list of nodes.
* Add a benchmark used to test checking how fast `GetNodes` is for a 4,000 node cluster.

**Related Issues**

Fixes https://github.com/gravitational/teleport/issues/1954

### CODE REVIEW

File: integration/helpers.go [hunk 1]
```diff
@@ -552,34 +552,26 @@ func (i *TeleInstance) CreateEx(trustedSecrets []*InstanceSecrets, tconf *servic
 }
 
 // StartNode starts a SSH node and connects it to the cluster.
-func (i *TeleInstance) StartNode(name string, sshPort int) (*service.TeleportProcess, error) {
+//func (i *TeleInstance) StartNode(name string, sshPort int) (*service.TeleportProcess, error) {
```
Comment (klizhentas) [inline integration/helpers.go]: Leftover comment?

File: lib/auth/auth_with_roles.go [hunk 2]
```diff
@@ -317,11 +330,73 @@ func (a *AuthWithRoles) UpsertNode(s services.Server) error {
 	return a.authServer.UpsertNode(s)
 }
 
+// filterNodes filters nodes based off the role of the logged in user.
+func (a *AuthWithRoles) filterNodes(nodes []services.Server) ([]services.Server, error) {
+	// For certain built-in roles, continue to allow access and return the full
+	// set of nodes to not break existing clusters during migration. This is
+	// also used by the proxy to cache a list of all nodes for it's smart
+	// resolution.
+	if a.hasBuiltinRole(string(teleport.RoleAdmin)) ||
```
Comment (klizhentas) [inline lib/auth/auth_with_roles.go]: Why this code even necessary? If using standard built in roles should just allow to execute this action in the code above, so we have one less code path? Also, why RemoteProxy can list all remote nodes?

File: lib/auth/auth_with_roles.go [hunk 3]
```diff
@@ -317,11 +330,73 @@ func (a *AuthWithRoles) UpsertNode(s services.Server) error {
...
+
+	// Filter nodes to return the ones for the connected identity.
+	startFilter := time.Now()
+	filteredNodes, err := a.filterNodes(nodes)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	elapsedFilter := time.Since(startFilter)
+
+	log.Debugf("GetServers(%v->%v) in %v for %v, %v to fetch, %v to filter.",
```
Comment (klizhentas) [inline lib/auth/auth_with_roles.go]: use structured logging here like in agentserver please to mark times

File: lib/services/role.go [hunk 4]
```diff
@@ -1543,41 +1547,52 @@ func (set RoleSet) CheckLoginDuration(ttl time.Duration) ([]string, error) {
...
 	var errs []error
 
-	// check deny rules first: a single matching namespace, label, or login from
+	// Check deny rules first: a single matching namespace, label, or login from
 	// the deny role set prohibits access.
 	for _, role := range set {
-		matchNamespace := MatchNamespace(role.GetNamespaces(Deny), s.GetNamespace())
-		matchLabels := MatchLabels(role.GetNodeLabels(Deny), s.GetAllLabels())
-		matchLogin := MatchLogin(role.GetLogins(Deny), login)
+		matchNamespace, namespaceError := MatchNamespace(role.GetNamespaces(Deny), s.GetNamespace())
```
Comment (klizhentas) [inline lib/services/role.go]: it's not an error, but rather debug message, let's rename

File: lib/services/role.go [hunk 5]
```diff
@@ -1543,41 +1547,52 @@ func (set RoleSet) CheckLoginDuration(ttl time.Duration) ([]string, error) {
...
+		matchLabels, labelError := MatchLabels(role.GetNodeLabels(Deny), s.GetAllLabels())
+		matchLogin, loginError := MatchLogin(role.GetLogins(Deny), login)
 		if matchNamespace && (matchLabels || matchLogin) {
-			errorMessage := fmt.Sprintf("role %v denied access to node %v: deny rule matched; match(namespace=%v, label=%v, login=%v)",
-				role.GetName(), s.GetHostname(), matchNamespace, matchLabels, matchLogin)
-			log.Warnf("[RBAC] Denied access to server: " + errorMessage)
-			return trace.AccessDenied(errorMessage)
+			if log.GetLevel() == log.DebugLevel {
```
Comment (klizhentas) [inline lib/services/role.go]: add comment here to explain why is it necessary

File: lib/services/role_test.go [hunk 6]
```diff
@@ -1330,3 +1332,65 @@ func (s *RoleSuite) TestCheckAndSetDefaults(c *C) {
 	}
 
 }
+
+func BenchmarkCheckAccessToServer(b *testing.B) {
```
Comment (klizhentas) [inline lib/services/role_test.go]: very nice, add comment though on how to run it with go test

File: lib/auth/auth_with_roles.go (hunk 2 above)
Comment (klizhentas) [reply lib/auth/auth_with_roles.go]: never mind this comment, I understand that it's a special case now, although I don't get why would RemoteProxy has this permission still?

Comment (russjones) [reply lib/auth/auth_with_roles.go]: For sure I know we have a test that uses `teleport.RoleRemoteProxy`. When I went to check where it was used with Trusted Clusters I found they were broken on master so I have not been able to check.

Will check this when Trusted Clusters are fixed.

File: lib/services/role.go (hunk 5 above)
Comment (russjones) [reply lib/services/role.go]: Since this is done in three places, I added the comment to the function explaining why logging is behind a debug flag.

**Review by klizhentas — APPROVED** [review]

**Pull request merged**